### PR TITLE
Spread checkpoint writing

### PR DESCRIPTION
**Pull request by codesome**
Due to a burst of disk writes during checkpoint, we were facing some heavy disk I/O throttling. This has a big effect on ingestion latency as we write to WAL (north of 60s in worst cases for around 500k series). Hence spread the writing of checkpoint to the entire checkpoint duration.

### CODE REVIEW

File: pkg/ingester/wal.go [hunk 1]
```diff
@@ -241,6 +241,24 @@ func (w *walWrapper) performCheckpoint() (err error) {
 		os.RemoveAll(checkpointDirTemp)
 	}()
 
+	// Count number of series - we'll use this to rate limit checkpoints.
+	numSeries := 0
+	for _, state := range w.getUserStates() {
```
Comment (pracucci) [inline pkg/ingester/wal.go]: `w.getUserStates()` does a copy of the map and it's also called below within the same file. You may just call it once and then reuse it.

File: pkg/ingester/wal.go [hunk 2]
```diff
@@ -241,6 +241,24 @@ func (w *walWrapper) performCheckpoint() (err error) {
...
+	}
+	if numSeries == 0 {
+		return nil
+	}
+
+	perSeriesDuration := w.cfg.CheckpointDuration / time.Duration(numSeries)
+	ticker := time.NewTicker(perSeriesDuration)
+
+	if immediate {
+		ticker.Stop()
```
Comment (pracucci) [inline pkg/ingester/wal.go]: What's the point of creating a `ticker` we immediately `Stop()`?

File: pkg/ingester/wal.go [hunk 3]
```diff
@@ -180,7 +180,7 @@ func (w *walWrapper) run() {
 
 const checkpointPrefix = "checkpoint."
 
-func (w *walWrapper) performCheckpoint() (err error) {
+func (w *walWrapper) performCheckpoint(immediate bool) (err error) {
```
Comment (pracucci) [inline pkg/ingester/wal.go]: With this change, there's no guarantee that the next `performCheckpoint()` (called by the priodically `CheckpointDuration` ticker) will be performed once the previous one has completed. Shouldn't we protect from this case?

File: pkg/ingester/wal.go (hunk 1 above)
Comment (codesome) [reply pkg/ingester/wal.go]: Right, modified

File: pkg/ingester/wal.go (hunk 2 above)
Comment (codesome) [reply pkg/ingester/wal.go]: I guess I was in a hurry :) I have modified it now to only initialize when needed

File: pkg/ingester/wal.go (hunk 3 above)
Comment (codesome) [reply pkg/ingester/wal.go]: Good catch! I thought it was protected by a mutex, but apparently now. I have added one now.

File: pkg/ingester/wal.go [hunk 4]
```diff
@@ -241,15 +245,40 @@ func (w *walWrapper) performCheckpoint() (err error) {
...
+	numSeries := 0
+	us := w.getUserStates()
+	for _, state := range us {
+		numSeries += state.fpToSeries.length()
+	}
+	if numSeries == 0 {
+		return nil
+	}
+
+	perSeriesDuration := w.cfg.CheckpointDuration / time.Duration(numSeries)
```
Comment (pracucci) [inline pkg/ingester/wal.go]: This can be moved inside the `if !immediate` block

File: pkg/ingester/wal.go [hunk 5]
```diff
@@ -180,11 +181,14 @@ func (w *walWrapper) run() {
 
 const checkpointPrefix = "checkpoint."
 
-func (w *walWrapper) performCheckpoint() (err error) {
+func (w *walWrapper) performCheckpoint(immediate bool) (err error) {
 	if !w.cfg.CheckpointEnabled {
 		return nil
 	}
 
+	w.checkpointMtx.Lock()
```
Comment (pracucci) [inline pkg/ingester/wal.go]: Please add a comment to explain why we do it. May not be much clear why we lock the entire function.

**Pull request merged**